src/net/ossl.cc

```diff
@@ -557,312 +607,560 @@ namespace tls {
+                            }
                         });
                     }
-                    _error = std::make_exception_ptr(ossl_error("Failed on call to SSL_write"));
-                    return make_exception_future<stop_iteration>(_error);
-                }
-                off += bytes_written;
-                /// Regardless of error, continue to send fragments
-                return pull_encrypted_and_send().then([]{
-                    return make_ready_future<stop_iteration>(stop_iteration::no);
+                ).then([&p] {
+                    return p.share();
                 });
-            });
-        });
+            }
+        );
     }
 
+    // Used to push unencrypted data through OpenSSL, which will
+    // encrypt it and then place it into the output bio.
     future<> put(net::packet p) override {
+        constexpr size_t openssl_max_record_size = 16 * 1024;
         if (_error) {
-            return make_exception_future<>(_error);
+            return make_exception_future(_error);
         }
         if (_shutdown) {
-            return make_exception_future<>(std::system_error(EPIPE, std::system_category()));
+            return make_exception_future<>(
+              std::system_error(EPIPE, std::system_category()));
         }
         if (!connected()) {
-            return handshake().then([this, p = std::move(p)]() mutable {
-               return put(std::move(p));
-            });
+            return handshake().then(
+              [this, p = std::move(p)]() mutable { return put(std::move(p)); });
         }
 
-        // We want to make sure that we write to the underlying bio with as large
-        // packets as possible. This is because eventually this translates to a
-        // sendmsg syscall. Further it results in larger TLS records which makes
-        // encryption/decryption faster. Hence to avoid cases where we would do
-        // an extra syscall for something like a 100 bytes header we linearize the
-        // packet if it's below the max TLS record size.
-        // TODO(Rob): Avoid magic numbers
-        if (p.nr_frags() > 1 && p.len() <= 16000) {
+        if (p.nr_frags() > 1 && p.len() <= openssl_max_record_size) {
             p.linearize();
         }
-
-        auto i = p.fragments().begin();
-        auto e = p.fragments().end();
-        return with_semaphore(_out_sem, 1, std::bind(&session::do_put, this, i, e)).finally([p = std::move(p)] {});
+        return get_units(_out_sem, 1).then([this, p = std::move(p)](auto units) mutable {
+            return do_put(std::move(p), std::move(units)).then([this](net::packet p) {
+                if (eof() || p.len() == 0) {
+                    // we have closed early, drop the packet and return
+                    return make_ready_future();
+                } else {
+                    return handshake().then([this, p = std::move(p)]() mutable {
+                        return put(std::move(p));
+                    });
```

Comment:
It's not clear to me that we need to call `handshake` here, and if we do, but we are connected() (SSL_is_init_finished), then we don't actually handle the WANT_READ/WRITE. Does `SSL_is_init_finished` return false during a renegotiation?

Reply:
> Does SSL_is_init_finished return false during a renegotiation?

yes

Reply:
I did test this logic by inducing a renegotiation (calling `SSL_renegotiate` once on the `put` path).